src/Tinfoil/AEAD/AESGCM/Constraints.hs

```diff
@@ -0,0 +1,83 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+
+{-
+This module contains constraints on message lengths and message counts
+which apply to AES256-GCM as implemented in tinfoil. The constraints
+are mostly the same as the recommendations in
+<dx.doi.org/10.6028/NIST.SP.800-38D NIST Special Publication 800-38D>;
+there are some additional constraints imposed by IV construction in our case.
+-}
+module Tinfoil.AEAD.AESGCM.Constraints (
+    Blocks(..)
+  , Bytes(..)
+  , Messages(..)
+
+  , blocksBytes
+  , maxCleartextLength
+  , maxMessageCount
+  , maxTotalData
+) where
+
+import           P
+
+-- |
+-- Number of blocks. A block is 128 bits.
+newtype Blocks = Blocks {
+    unBlocks :: Integer
+  } deriving (Eq, Show, Ord, Enum, Num)
+
+-- |
+-- Number of bytes. Assumes that bytes are equivalent to octets; this is
+-- true on all supported architectures (i.e., x86_64).
+newtype Bytes = Bytes {
+    unBytes :: Integer
+  } deriving (Eq, Show, Ord, Enum, Num)
+
+-- |
+-- Count (number) of messages.
+newtype Messages = Messages {
+    unMessages :: Integer
+  } deriving (Eq, Show, Ord, Enum, Num)
+
+-- |
+-- Converts blocks to bytes (octets).
+blocksBytes :: Blocks -> Bytes
+blocksBytes (Blocks n) =
+  Bytes $ (n * 128) `div` 8
+
+-- |
+-- Upper bound on octets encrypted with one (key, IV) pair (i.e., one
+-- message). This is around 64GiB.
+--
+-- GCM's internal counter is 32 bits, so we can't encrypt more than
+-- @2^32@ blocks; block size is @2^7@ bits, so we have @2^39@ bits worth of
+-- counter. We lose one block due to the GCM counter skipping 0, and another
+-- to encrypt the @GHASH@ with the first block in the keystream.
+--
+-- So in blocks, we have a limit of @(2^32 - 2)@.
+--
+-- <dx.doi.org/10.6028/NIST.SP.800-38D NIST SP 800-38D>
+maxCleartextLength :: Blocks
+maxCleartextLength =
+  Blocks 4294967294
```

Comment:
super easy to find out `mafia build --dump-core` then look in `dist/build/src/Tinfoil/AEAD/AESGCM/Constraints.dump-simpl`